content/scripts/build.js

```diff
@@ -414,23 +395,21 @@ class Builder {
     for (const preUri of prerequisites) {
       const preCleanUri = this.cleanUri(preUri);
       if (this.allTitles.has(preCleanUri)) {
-        const preFolder = this.getFolder(preCleanUri);
-        const preSource = this.getSource(preFolder);
-        // TODO: What if this prerequisite is a stumptown document?
-        //       Probably should be an error that we report?
-        if (!preSource.isStumptown) {
-          const preMetadata = getMetadata(preSource, preFolder).metadata;
-          const preRawHtml = fs.readFileSync(
-            path.join(preFolder, "index.html"),
-            "utf8"
-          );
-          // When rendering prerequisites, we're only interested in
-          // caching the results for later use. We don't care about
-          // the results returned.
-          await this.renderMacros(preRawHtml, preMetadata, {
-            cacheResult: true,
-          });
+        const folder = this.allTitles.get(preCleanUri).file;
+        if (source.isStumptown) {
+          throw new Error("Stumptown sources don't have macros.");
         }
```

Comment:
Since we're reusing the incoming `source` for all of the prerequisites, these lines can be deleted. The error will never be thrown, since the `renderMacros()` method is not, and should never be, called for "stumptown" sources.

This check was originally present in the code because a prerequisite document is always a different document from the incoming document, and therefore _might_ be from a different source, assuming of course that we allow a mixture of stumptown and macro-based documents within the same build. That's why even though I like the simplicity of this PR, I don't like that it assumes that all of the documents in the prerequisite chain are from the same source. Even though that may be true now, it may not be true in the future when we allow a mixture of stumptown and macro-based documents. Also, even if we suspect that in the future the mixture will be such that no macro-based document will ever have a stumptown document in its prerequisite chain, we should probably keep the original code here so we can check that and throw an error if that doesn't hold true.

However, if our working assumption is that we will **not** ever allow a mixture of stumptown and macro-based documents, that any build must be one or the other, then this PR is awesome.

Reply:
@peterbe I woke up, couldn't sleep, and then realized while laying awake that this change still doesn't solve the fundamental problem that forced me to add `getFolder()`, and that's the "server" case. The problem only occurs when running `yarn start`. Since `this.allTitles.get(preCleanUri).file` always returns a directory relative to the repo root, this line:
https://github.com/mdn/stumptown-renderer/blob/900b2f888f07d3b67feb1f226551aa649adfbac3/content/scripts/build.js#L143
will always fail in that case because the folder is relative to `process.cwd()`, which isn't the repo root as it is when running `node content build ...` but rather `<root>/server` because of the way the server is started, and so the directory/file doesn't exist. For example, it'll look for the file in `ROOT/server/content/files/en-us/...` rather than `ROOT/content/files/en-us/...`.

So, in other words, apart from our source discussion above, which is important too, this change still won't work because the working directory of the builder when running `yarn start` is `ROOT/server` rather than `ROOT`, and since all of the paths in `allTitles` are relative to `ROOT`, they have to be adjusted when running from the server.

Another way to tackle that problem, apart from my `getFolder()` approach, would be to ensure that the server adjusts its working directory to `ROOT` after starting (and adjusts whatever paths need to be adjusted for itself, for example it would no longer need to use `normalizeContentPath()` when specifying sources for the builder since the directories relative to `ROOT` would work fine). I kind of like the idea that whatever command we run must run from `ROOT` (i.e., `process.cwd()` would always return `ROOT`), so that we can always use relative folders/filepaths.

Reply:
To test what I'm saying, do a `yarn clean && yarn start` and then go to a page that has at least one prerequisite, for example http://localhost:3000/en-us/docs/web/accessibility/aria/roles/alert_role and you'll see this error:

```sh
1:24:09 AM web.1          |  Files successfully emitted, waiting for typecheck results...
1:24:09 AM web.1          |  Compiled successfully!
1:24:09 AM web.1          |  You can now view client in the browser.
1:24:09 AM web.1          |    Local:            http://localhost:3000
1:24:09 AM web.1          |    On Your Network:  http://192.168.1.18:3000
1:24:09 AM web.1          |  Note that the development build is not optimized.
1:24:09 AM web.1          |  To create a production build, use npm run build.
1:25:18 AM server.1       |  Building a map of ALL redirects...
1:25:18 AM server.1       |  Building map of all redirects took 86.0 milliseconds
1:25:18 AM server.1       |  Error while processing: /Users/rjohnson/repos/yari/content/files/en-us/web/accessibility/aria/roles/alert_role
1:25:18 AM server.1       |  Error: ENOENT: no such file or directory, open 'content/files/en-us/web/accessibility/aria/roles/index.yaml'
1:25:18 AM server.1       |      at Object.openSync (fs.js:440:3)
1:25:18 AM server.1       |      at Object.readFileSync (fs.js:342:35)
1:25:18 AM server.1       |      at getMetadata (/Users/rjohnson/repos/yari/content/scripts/build.js:141:26)
1:25:18 AM server.1       |      at Builder.renderMacros (/Users/rjohnson/repos/yari/content/scripts/build.js:402:29)
1:25:18 AM server.1       |      at Builder.processFolder (/Users/rjohnson/repos/yari/content/scripts/build.js:1148:42)
1:25:18 AM server.1       |      at Builder.start (/Users/rjohnson/repos/yari/content/scripts/build.js:451:34)
1:25:18 AM server.1       |      at /Users/rjohnson/repos/yari/server/index.js:179:50
1:25:18 AM server.1       |      at Layer.handle [as handle_request] (/Users/rjohnson/repos/yari/node_modules/express/lib/router/layer.js:95:5)
1:25:18 AM server.1       |      at next (/Users/rjohnson/repos/yari/node_modules/express/lib/router/route.js:137:13)
1:25:18 AM server.1       |      at Route.dispatch (/Users/rjohnson/repos/yari/node_modules/express/lib/router/route.js:112:3) {
1:25:18 AM server.1       |    errno: -2,
1:25:18 AM server.1       |    syscall: 'open',
1:25:18 AM server.1       |    code: 'ENOENT',
1:25:18 AM server.1       |    path: 'content/files/en-us/web/accessibility/aria/roles/index.yaml'
1:25:18 AM server.1       |  }
```

Reply:
The `server` does everything in absolute URLs. I.e. it creates `Builder` instance with things like `/Users/peterbe/dev/MOZILLA/MDN/stumptown-renderer/content/files` and not `content/files`. 
Actually, all the problems in the world is solved if you do two things:

1. Make the `.file` value in `_all-title.json` absolute. 
2. Fix the `normalizeContentPath` so it's able to cope with already absolute paths. 

I have a pending fix.

Reply:
The above example of breakage no longer happens. All is well now. 
Absolute full paths are making me sleep better at night and the only time they're "bad" is if they leak outside the internals. For example, you wouldn't want to ship `index.json` to S3 that contains something like `/home/travis/yari/client/build...` But that's not happening anywhere as far as I can see.